### PR TITLE
AC-12731 Issue#39020 Rendering the critical css inline event via csp …

### LINKED ISSUE #39020: CSP issues combined with dev/css/use_css_critical_path
### Preconditions and environment

- Magento version: 2.4.7-p1

### Steps to reproduce

1. Enable `dev/css/use_css_critical_path` in order to facilitate critical CSS functionality
2. Add some items to cart
3. Visit checkout page

### Expected result

Checkout pages are rendered with proper css styles

### Actual result

Restricted CSP policy (introduced in 2.4.7 for checkout pages) prevents inline js to be executed so CSS files which were supposed to be loaded asynchronously are never loaded.

Dev Console gives such output:
Screenshot 2024-08-07 at 14 50 21

For following elements:
Screenshot 2024-08-07 at 14 55 20

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [X] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Mohamed-Asar**
### Description (*)
Rendering the critical css inline event via csp inline secure render

### Fixed Issues (if relevant)

1. Fixes magento/magento2#39020

### Manual testing scenarios (*)

1. Enable dev/css/use_css_critical_path in order to facilitate critical CSS functionality
2. Add some items to cart
3. Visit checkout page

### CODE REVIEW

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php [hunk 1]
```diff
@@ -26,12 +27,21 @@ class AsyncCssPlugin
...
+
     /**
      * @param ScopeConfigInterface $scopeConfig
+     * @param InlineUtilInterface $cspInlineUtil
      */
-    public function __construct(ScopeConfigInterface $scopeConfig)
-    {
+    public function __construct(
+        ScopeConfigInterface $scopeConfig,
+        InlineUtilInterface $cspInlineUtil
```
Comment (engcom-Hotel) [inline app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php]: New Params in Constructor should be null and handle it with ObjectManager

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @Mohamed-Asar,

Thanks for the contribution!

Please make the changes as suggested below.

Thanks

File: app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php (hunk 1 above)
Comment (Mohamed-Asar) [reply app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php]: @engcom-Hotel  Changes done. please check

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
It seems that review changes has been covered so moved it back to review

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests CE, Functional Tests EE, Static Tests, Unit Tests

**Comment by engcom-Bravo** [comment]
Hi @Mohamed-Asar,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**


- Enable dev/css/use_css_critical_path in order to facilitate critical CSS functionality
- Add some items to cart
- Visit checkout page


**Before:** :heavy_multiplication_x: 



Screenshot 2024-09-23 at 11 00 33



**After:** :heavy_check_mark: 


Screenshot 2024-09-23 at 11 04 54


Checkout pages are rendered with proper css styles

**Comment by engcom-Bravo** [comment]
@magento run all tests

**Comment by engcom-Bravo** [comment]
Hi @Mohamed-Asar,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**


- Enable dev/css/use_css_critical_path in order to facilitate critical CSS functionality
- Add some items to cart
- Visit checkout page


**Before:** :heavy_multiplication_x: 



Screenshot 2024-09-23 at 11 00 33



**After:** :heavy_check_mark: 


Screenshot 2024-09-23 at 11 04 54


Checkout pages are rendered with proper css styles



Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Static Tests, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Static Tests, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Integration Tests, Static Tests, Unit Tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@engcom-Charlie I fixed the unit cases issue

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by Mohamed-Asar** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

**Comment by Mohamed-Asar** [comment]
@magento run Static Tests

**Comment by Mohamed-Asar** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B, Functional Tests EE, Static Tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE,  Functional Tests B2B

**Comment by engcom-Charlie** [comment]
The functional B2B failure are not consistent in recent 2 builds. They are neither failing because of PR changes nor part of it. Considering them flaky.

**Run 1:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39040/19e86db46db2fb9a04191525e3bca7ec/Functional/allure-report-b2b/index.html#categories

image


**Run 2:**
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39040/67cd301cc12b89e8cc9cff6f7b6a2aa6/Functional/allure-report-b2b/index.html#categories

image

**Comment by bigbangx** [comment]
Hello,
Not sure if that's what you expected, but it's good to mention that, as `$useUnsafeHashes` is `false` by defalut in `Magento\Csp\Helper\InlineUtil` , current code will set  `unsafe-inline` on script-src for all pages with critical css !

**Pull request merged**